### PR TITLE
Add creation steps for smart poster, local and external type embedded

### PR DESCRIPTION
**Pull request by zolkis**
***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/web-nfc/pull/495.html" title="Last updated on Jan 2, 2020, 1:28 PM UTC (7374a39)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-nfc/495/5a3d580...zolkis:7374a39.html" title="Last updated on Jan 2, 2020, 1:28 PM UTC (7374a39)">Diff</a>

### CODE REVIEW

**Comment by zolkis** [comment]
Added the local + external type handling (including embedded messages) to this PR since they are both connected to #475.

**Comment by zolkis** [comment]
@leonhsl could you please review this, too?

File: index.html [hunk 1]
```diff
@@ -3303,32 +3311,86 @@ <h3><dfn>Writing or pushing content</dfn></h3>
...
+              </li>
+              <li>
+                Set |data:byte sequence| to |arrayBuffer|.[[\ArrayBufferData]].
+              </li>
+              <li>
+                Set the |ndef|'s <a>PAYLOAD LENGTH field</a> to |length|.
+              </li>
+              <li>
+                If |length| > `0`, set the |ndef|'s <a>PAYLOAD field</a>
+                to |data|.
```
Comment (kenchris) [inline index.html:3342]: Is data initialized to null?

File: index.html [hunk 2]
```diff
@@ -3303,32 +3311,86 @@ <h3><dfn>Writing or pushing content</dfn></h3>
...
+            Return |ndef|.
+          </li>
+        </ol>
+      </div>
+    </section>
+
+    <section><h3>Mapping smart poster to NDEF</h3>
+      <div>
+        To <dfn>map smart poster to NDEF</dfn>, given a |record:NDEFRecordInit|
+        and |ndef|,run these steps:
```
Comment (kenchris) [inline index.html]: Missing space after ,

File: index.html (hunk 1 above)
Comment (zolkis) [reply index.html:3342]: No, the algorithms builds it, starting from an empty byte sequence.

File: index.html [hunk 3]
```diff
@@ -3303,32 +3311,86 @@ <h3><dfn>Writing or pushing content</dfn></h3>
           </li>
           <li>
             If the type of a |record|'s <a>data</a> is not a
-            {{BufferSource}}, [= exception/throw =] a {{TypeError}}
-            and abort these steps.
+            {{BufferSource}} or {{NDEFMessageInit}}, [= exception/throw =] a
```
Comment (beaufortfrancois) [inline index.html:3314]: As "sub records" are allowed for a local type record, we need to adjust https://pr-preview.s3.amazonaws.com/zolkis/web-nfc/pull/495.html#data-mapping as well.

File: index.html [hunk 4]
```diff
@@ -3279,11 +3260,38 @@ <h3><dfn>Writing or pushing content</dfn></h3>
             <a>recordType</a>.
           </li>
           <li>
-            Set the |ndef|'s <a>PAYLOAD LENGTH field</a> to |length|.
+            If the type of a |record|'s <a>data</a> is a {{BufferSource}},
```
Comment (beaufortfrancois) [inline index.html:3263]: It looks like steps 3 and 4 are the only differences between "Mapping external data to NDEF" and "Mapping local type to NDEF". Is there a way to emphasize this?

File: index.html [hunk 5]
```diff
@@ -3303,32 +3311,86 @@ <h3><dfn>Writing or pushing content</dfn></h3>
...
             ([=well-known type record=]).
           </li>
           <li>
-            Set |ndef|'s <a>TYPE field</a> to |record|'s
-            <a>recordType</a>, representing the <a>local type</a> name.
+            Set |ndef|'s <a>TYPE field</a> to "`Sp`" (`0x53` `0x70`).
           </li>
           <li>
-            Set the |ndef|'s <a>PAYLOAD LENGTH field</a> to |length|.
+            set |ndef|'s <a>PAYLOAD field</a> to the result of running the
```
Comment (beaufortfrancois) [inline index.html]: Nit: `s/set/Set/`

**Review by beaufortfrancois — APPROVED** [review]
LGTM with nits

File: index.html (hunk 4 above)
Comment (zolkis) [reply index.html:3263]: Good observation. Let's factor out that as a parametrised algorithm in a next PR. Should we track this as an issue?

**Comment by zolkis** [comment]
Fixed the nits, could we squash+merge?

File: index.html (hunk 4 above)
Comment (beaufortfrancois) [reply index.html:3263]: Sure. Please open one and link to this comment.

**Comment by beaufortfrancois** [comment]
I'm good.
@kenchris for thoughts as well.

**Review by kenchris — APPROVED** [review]

**Comment by leonhsl** [comment]
lgtm
FYI, `absolute-url` record type is already fully implemented, you might also want to add it into the spec.

**Comment by beaufortfrancois** [comment]
@leonhsl `absolute-url` records are already in the spec at https://w3c.github.io/web-nfc/#parsing-ndef-absolute-url-records. However, it looks like the "create NDEF record" algorithm is missing some parts: https://w3c.github.io/web-nfc/#creating-ndef-record
@zolkis Do you want to tackle this?

**Comment by zolkis** [comment]
Yessir, will check if we have an issue for this, if not, I file and handle it.

**Pull request merged**